### PR TITLE
addVM, editVM dialogs

### PR DESCRIPTION
**Pull request by mkrajnak**
Provides `Create VM` and `Edit VM` functionality.

In addition to review comments in the code:
- [x] move the "vm name" under the "os type"
- [x] change the memory to be in MiB (and the label to contain that it is in MiB)
- [x] change the combo boxes to be type aheads (https://www.patternfly.org/pattern-library/widgets/#bootstrap-combobox)
- [x] sort the date in combo boxes
- [x] after submit close the dialog
- [x] fail handling - reasonable user notification
- [x] Step for memory increase/decrease by Up/Down buttons should be 256MB

---
Will fix https://github.com/oVirt/ovirt-web-ui/issues/17

### CODE REVIEW

**Comment by jelkosz** [comment]
@vojtechszocs FYI

**Comment by jelkosz** [comment]
couple of notes:

- [ ] move the "vm name" under the "os type"
- [ ] change the memory to be in MiB (and the label to contain that it is in MiB)
- [ ] change the combo boxes to be type aheads (https://www.patternfly.org/pattern-library/widgets/#bootstrap-combobox)
- [ ] sort the date in combo boxes
- [ ] after submit close the dialog

**Comment by mkrajnak** [comment]
@jelkosz All Done, except 

> change the combo boxes to be type aheads (https://www.patternfly.org/pattern-library/widgets/#bootstrap-combobox)

There is one problem which I am not sure how to solve properly. When I install bootstrap-combobox via yarn or npm installed package is this https://github.com/cchantep/bootstrap-combobox. It lacks the functionality we want.
Combobox used by patterfly is this one https://github.com/danielfarrell/bootstrap-combobox. I tried to install it via yarn from github, which should work as it is mentioned in documentation but it ends with error. So I cloned it manually to node_modules, which temporarily solved the problem. 
How should I properly include right package ?

**Comment by michalskrivanek** [comment]
opened https://ovirt-jira.atlassian.net/browse/OVIRT-1249 to stop spam from lago-bot

**Comment by landgraf** [comment]
ci test please

File: src/sagas.js [hunk 1]
```diff
@@ -23,7 +23,41 @@ import {
 } from 'ovirt-ui-components'
 
 // import store from './store'
-import { persistState, getSingleVm } from './actions'
+import {
+  persistState,
```
Comment (matobet) [inline src/sagas.js]: I would vote for moving all *VmDialog related sagas to a separate file. This is starting to become unmanageable.

File: src/reducers/reducers.js [hunk 2]
```diff
@@ -8,6 +8,13 @@ import {
   VisibilityReducer as visibility,
 } from 'ovirt-ui-components'
```
Comment (matobet) [inline src/reducers/reducers.js]: when moving from a single file `foo.js` to a directory `foo` with mutliple files I would prefer the main file to be called `index.js`. This will allow other code that imported `foo` to work as before and also enforces standardized naming across the modules.

File: src/sagas.js (hunk 1 above)
Comment (mareklibra) [inline src/sagas.js]: Agree

**Comment by landgraf** [comment]
ci test please

File: src/reducers/templates.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import Immutable from 'immutable'
+import { actionReducer } from './utils'
+
+const templates = actionReducer(Immutable.fromJS({ templates: {}, loadInProgress: true }), {
+  ADD_TEMPLATES (state, { payload: { templates } }) {
+    const updates = {}
```
Comment (matobet) [inline src/reducers/templates.js]: Could you please wrap this peace of code into some sort of `groupById` utility function? I believe I saw this piece of code copy repeated in many places.

File: src/reducers/vmDialog.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  cluster: {},
+  template: {},
+  os: {},
+  name: '',
+  memory: '',
+  cpu: {},
+  errorMessage: '',
+}), {
+  UPDATE_CLUSTER (state, { payload: { cluster } }) {
+    const imUpdates = Immutable.fromJS(cluster)
```
Comment (matobet) [inline src/reducers/vmDialog.js]: How about inlining those trivial intermediate variables? If after inlining it would appear too long you can also consider importing the `fromJS` function using a named import: `import { fromJS } from 'immutable'`.

File: src/selectors.js [hunk 5]
```diff
@@ -17,6 +17,21 @@ Selectors = {
   getVmDisks ({ vmId }) {
     return getState().vms.getIn(['vms', vmId, 'disks'])
   },
+  getFirstCluster () {
+    return getState().clusters.get('clusters').toList().first()
+  },
+  getClusterById (id) {
+    return getState().clusters.get('clusters').toList().find(cluster =>
```
Comment (matobet) [inline src/selectors.js]: If I read the reducers correctly, those entities (clusters/templates/oss) are already stored in ID-indexed hash-maps so I really not see the reason behind doing this `O(n)` lookup when a simple get by id from the immutable map should suffice.

File: src/templateDialog.js [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+  updateEditTemplateDescription,
+  updateEditTemplateOS,
+  updateEditTemplateMemory,
+  updateEditTemplateCpu,
+  changeEditTemplate,
+  editTemplate,
+  closeDetail,
+} from './actions'
+
+class templateDialog extends React.Component {
```
Comment (matobet) [inline src/templateDialog.js]: components with `PascalCase` please - `TemplateDialog`

File: src/templateDialog.js [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    this.props.changeName(this.templateName.value)
+  }
+
+  getOperatingSystem (name) {
+    return this.props.operatingSystems.get('operatingSystems').toList().find(os =>
+      os.get('description') === name)
+  }
+
+  getTemplate (name) {
+    return this.props.templates.get('templates').toList().find(template =>
```
Comment (matobet) [inline src/templateDialog.js]: selector wanting to be extracted?: e.g. `getTemplateByName`

File: src/templateDialog.js [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    $(this.template).selectpicker()
+    $(this.templateOs).selectpicker()
+  }
+
+  changeName () {
+    this.props.changeName(this.templateName.value)
+  }
+
+  getOperatingSystem (name) {
+    return this.props.operatingSystems.get('operatingSystems').toList().find(os =>
```
Comment (matobet) [inline src/templateDialog.js]: same as below

File: src/templateDialog.js [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+
+  render () {
+    return (
+      <DetailContainer>
+        <h1>Edit Template</h1>
+        <hr />
+        <ErrorAlert message={this.props.errorMessage} />
+        <form className='form-horizontal'>
+          <LabeledSelect
+            label='Templates'
```
Comment (matobet) [inline src/templateDialog.js]: @mareklibra please what is the current strategy regarding i18n in ovirt-web-ui? Should we already in this PR for the dialog code use some sort of i18n infra?

File: src/labeledTextField.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  type: PropTypes.string,
+  key: PropTypes.string,
+  value: PropTypes.oneOfType([
+    PropTypes.string,
+    PropTypes.number,
+  ]),
+  disabled: PropTypes.bool,
+}
+
+export default connect()(LabeledTextField)
```
Comment (matobet) [inline src/labeledTextField.js]: Does this component need to be connected to Redux at all? To me it seems as a pure presentational ("dumb") component only depending on props.

File: src/templateDialog.js (hunk 9 above)
Comment (gregsheremeta) [inline src/templateDialog.js]: feel free to ping @vojtechszocs and @sjd78 for their thoughts on i18n, especially if you're going to hook into Zanata like ovirt-engine-dashboard does.

File: src/reducers/dialogVisibility.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
+import Immutable from 'immutable'
+import { actionReducer } from './utils'
+
+const dialogVisibility = actionReducer(Immutable.fromJS({ showVmDialog: false, showVmDetail: false, showEditTemplate: false }), {
```
Comment (matobet) [inline src/reducers/dialogVisibility.js]: Now considering how this will scale for the future when we will have even more dialog types I believe a bit different store representation would be more suitable. 

One alternative would be to have a simple `dialog` field there with a string identifying the type of dialog that is to be shown (since I assume this sort of "top level" dialog will be possible to display only one at a time) with values like (`null`, `"VM_EDIT"`, `"TEMPLATE_EDIT"`, ...)

Other to have this split by 2 levels: one field identifying the general dialog and second the current "behavior" (sorry for the pun with legacy gwt UICommon model :-)) or it. But in any case I would prefer to avoid this explosion of boolean flags.....

What do you guys think: @jelkosz @mareklibra @vojtechszocs ?

File: src/reducers/editTemplate.js [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+  name: '',
+  memory: '',
+  cpu: '',
+  errorMessage: '',
+}), {
+  UPDATE_EDIT_TEMPLATE (state, { payload: { template } }) {
+    return Immutable.fromJS(template)
+  },
+  UPDATE_EDIT_TEMPLATE_NAME (state, { payload: { name } }) {
+    const imUpdates = Immutable.fromJS(name)
```
Comment (matobet) [inline src/reducers/editTemplate.js]: same comment about removing intermediate variables

File: src/reducers/dialogVisibility.js (hunk 11 above)
Comment (jelkosz) [inline src/reducers/dialogVisibility.js]: yes, this representation is a call for trouble because it is a question of time until due to some bug you will have {showVmDialog: true, showVmDetail: true, showEditTemplate: true} - and now what :)

The struct: {type: ..., behavior: ...} sounds good to me. Could be: {type: VM, behavior: edit}

File: src/templateDialog.js (hunk 9 above)
Comment (mareklibra) [inline src/templateDialog.js]: I18n will be added to the project as a separate PR, I'm considering gettext to extract the strings.

File: src/index.js [hunk 13]
```diff
@@ -13,6 +13,10 @@ window.$ = window.jQuery = require('jquery')
 require('bootstrap/dist/js/bootstrap')
 window.patternfly = {}
 window.patternfly = require('patternfly/dist/js/patternfly')
+window.selectpicker = {}
+window.selectpicker = require('bootstrap-select/js/bootstrap-select.js')
```
Comment (mareklibra) [inline src/index.js]: @mkrajnak , please add bootstrap-select to package.json

File: src/index.js [hunk 14]
```diff
@@ -13,6 +13,10 @@ window.$ = window.jQuery = require('jquery')
 require('bootstrap/dist/js/bootstrap')
 window.patternfly = {}
 window.patternfly = require('patternfly/dist/js/patternfly')
+window.selectpicker = {}
```
Comment (mareklibra) [inline src/index.js]: I would need to retry, but as IIRC it's a hack to allow lazy init of those libraries.
I would not consider it as a pattern, so for the selectpicker please first try without it. If it works, let's do not use it.

File: src/reducers/dialogVisibility.js (hunk 11 above)
Comment (mkrajnak) [inline src/reducers/dialogVisibility.js]: @mbetak @jelkosz This is basically a hack that I am able to properly handle in sagas.

>  {showVmDialog: true, showVmDetail: true, showEditTemplate: true}
This is actually necessary because the only way to "Open the Doors" is to dispatch showVmDetail as shown bellow, to open addVmDialog:
`function* showAddNewVm (action) {
  yield put(setVmDetailToShow({ vmId: '0' })) // open the gate
 .......... 
 yield put(openVmDialog())
}`
 I know it is not a best solution,  @mbetak has a good suggestion I can try it later.

Comment (mareklibra) [inline src/reducers/dialogVisibility.js]: Somewhere else, I'm using this structure:
{
topLevel: 'type as String constant  or Object with details'
// if really required later, as a hack we can add another key along the topLevel, it would be still manageble
type: {
 // type specific map of details
 // if complex, then tree structure as the topLevel
 // but oVirt web admin is a good example that there's no need for very complex 2nd or 3rd level 
}
}

The reudcer (read: at one place) handles potential complexities of dependencies.

It works well. The ImmutableJS is a big benefit.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

File: src/TemplateDialog.js [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+        'type': this.templateOs.value,
+      },
+    }
+    this.props.editTemplate(template, this.props.template.get('id'))
+  }
+
+  render () {
+    return (
+      <DetailContainer>
+        <h1>Edit Template</h1>
```
Comment (jelkosz) [inline src/TemplateDialog.js]: This whole thing is from large part a copy/paste of the VmDialog. There are couple of issues with this, namely:
- when this becomes bigger, it becomes a maintenance nightmare
- when this will edit more entities (pools, instance types) it will become a maintenance nightmare on steroids (trust me, we had a similar issue with this dialog on the ovirt and we were drowning in in bugs like "this field is working in edit VM and Pool, but does not work in instance type" etc.)
- even now when it is small you have a bug like this here :) The memory label in VM has the changed label "Memory (MB)" and here it is still "Memory".

File: src/sagas.js [hunk 16]
```diff
@@ -245,6 +282,80 @@ function* startVm (action) {
...
+}
+
+function* handleClusterChange (action) {
+  yield put(updateCluster(action.payload.cluster))
+  // After every cluster change, set template to Blank
+  const blankTemplate = Selectors.getTemplateById('00000000-0000-0000-0000-000000000000')
+  yield put(changeTemplate(blankTemplate))
+}
+
+function* handleTemplateChange (action) {
```
Comment (jelkosz) [inline src/sagas.js]: It would be maybe good to do this right now while the patch is still young and the dialog does not have too much properties. 
So, what I mean - imagine this scenario:
- you have a VM and it was a hard labor to get all it's properties tuned according to your need.
- it is working and all is fine
- than you get access to a new cluster with more powerful hosts or whatever so you want to move it there
- so, you open the edit VM dialog, change the cluster and all is gone. All your settings are reset to defaults

Here we need to be a bit more smart. Either distinguish between new and edit (what I don't like but it is what we do in oVirt) or be even more smart and be able to decide if this field has been modified by the user or did not.
For edit VM in initialization all fields would be marked as modified by user and for new none. Than if you touch on new some field, it will be marked as modified by user. Than, if you change the template for example, the particular value will be rewritten from template only if it does not make sense in the context of new template, otherwise will stay intact.

@mkrajnak @mareklibra @jniederm @vojtechszocs @matobet what you guys think?

File: src/TemplateDialog.js (hunk 15 above)
Comment (mkrajnak) [inline src/TemplateDialog.js]: I think I will be able to wrap Operating system, memory, cpus and template to separated component.

File: src/sagas.js (hunk 16 above)
Comment (jelkosz) [inline src/sagas.js]: OK, it may turn out to be a pretty complicated change - can be done later. This patch brings lots of value also without this proposed change.

File: src/TemplateDialog.js (hunk 15 above)
Comment (jelkosz) [inline src/TemplateDialog.js]: yes, that sounds good

**Comment by mareklibra** [comment]
Rebase is needed because of https://github.com/oVirt/ovirt-web-ui/pull/109

**Comment by mkrajnak** [comment]
I pushed initial rebase

File: src/sagas.js [hunk 17]
```diff
@@ -295,19 +402,106 @@ function* schedulerPerMinute (action) {
...
+}
+
+function* fetchAllOS (action) {
+  const operatingSystems = yield callExternalAction('getAllOperatingSystems', Api.getAllOperatingSystems, action)
+
+  if (operatingSystems && operatingSystems['operating_system']) {
+    const operatingSystemsInternal = operatingSystems.operating_system.map(os => Api.OSToInternal({ os }))
+    yield put(addAllOS({ os: operatingSystemsInternal }))
+    const activeOperatingSystem = operatingSystemsInternal.find(os => os.name === 'other')
+    yield put(updateOperatingSystem(activeOperatingSystem))
```
Comment (mareklibra) [inline src/sagas.js]: This flow seems weird to me, please describe what's happening here.
Do I understand correctly, it's "updating a copy" held by potentially opened dialog? If so, the flow should be redesigned. @matobet , what do you think?

File: src/sagas.js [hunk 18]
```diff
@@ -295,19 +402,106 @@ function* schedulerPerMinute (action) {
...
+  }
+}
+
+function* fetchAllOS (action) {
+  const operatingSystems = yield callExternalAction('getAllOperatingSystems', Api.getAllOperatingSystems, action)
+
+  if (operatingSystems && operatingSystems['operating_system']) {
+    const operatingSystemsInternal = operatingSystems.operating_system.map(os => Api.OSToInternal({ os }))
+    yield put(addAllOS({ os: operatingSystemsInternal }))
+    const activeOperatingSystem = operatingSystemsInternal.find(os => os.name === 'other')
```
Comment (mareklibra) [inline src/sagas.js]: Why this 'other' constant?

File: src/actions/dialog.js [hunk 19]
```diff
@@ -0,0 +1,304 @@
+export function updateVmDialogErrorMessage (message) {
```
Comment (mareklibra) [inline src/actions/dialog.js]: Please rename this file to something more descriptive and related to its purpose.

File: src/sagas.js [hunk 20]
```diff
@@ -245,6 +277,80 @@ function* startVm (action) {
   yield stopProgress({ vmId: action.payload.vmId, name: 'start', result })
 }
 
+function* showEditVm (action) {
+  yield put(updateDialogType('edit'))
```
Comment (mareklibra) [inline src/sagas.js]: Maybe I'm missing something, but I don't prefer saga to be used for opening a dialog.
It seems to be too procedural.

Much better approach for Edit VM Dialog can be via React itself, something like following pseudo-code:
  <EditVM vm={myVmToEdit} clusters={clusters} templates={allTemplates} ...  />

The EditVm component state consists of a VM described as references to other resources, similar to:
this.state = {
vm: {
name,
cpus,
clusterId,
templateId,
...
}
}

With this approach, we
- avoid of this procedural code
- leverage React/Redux for updates if underlying sub-resources are changed or gradually loaded
- move rendering-related code to React component

@jelkosz , @matobet , please comment.

File: src/sagas.js [hunk 21]
```diff
@@ -295,19 +402,106 @@ function* schedulerPerMinute (action) {
   }
 }
 
+function* createNewVm (action) {
```
Comment (mareklibra) [inline src/sagas.js]: again, I think it's much better to wrap this logic into a React component and dispatch a "CreateVM" action once the "Submit" button is clicked.

**Comment by mareklibra** [comment]
Regarding comboboxes (see review), this shall be more elegant: http://www.patternfly.org/pattern-library/widgets/#bootstrap-combobox

**Comment by mkrajnak** [comment]
@mareklibra its the same bootstrap-combobox like on page you posted, X will appear when you have actual value inside
image

**Comment by mareklibra** [comment]
I missed that, thanks.

No matter it's patternfly, I don't like this in our use case. 
What about preselecting a default value (means "Blank" for template, "default" for cluster) if none is selected?
We would not deviate from the patternfly behavior much and the dialog would remain in consistent state.

@jelkosz , @matobet , @mkrajnak , what do you think?

**Comment by jelkosz** [comment]
@mareklibra yes, makes sense

**Comment by mareklibra** [comment]
WIP: follow-up patches to resolve review issues above

File: src/ovirtapi.js [hunk 22]
```diff
@@ -184,21 +250,48 @@ OvirtApi = {
...
+    return OvirtApi._httpGet({ url })
+  },
+  getAllClusters () {
+    OvirtApi._assertLogin({ methodName: 'getAllClusters' })
+    const url = '/api/clusters'
+    return OvirtApi._httpGet({ url })
+  },
+  getAllOperatingSystems () {
+    OvirtApi._assertLogin({ methodName: 'getAllOperatingSystems' })
+    const url = '/api/operatingsystems'
```
Comment (mareklibra) [inline src/ovirtapi.js]: Done

File: src/actions/dialog.js (hunk 19 above)
Comment (mareklibra) [inline src/actions/dialog.js]: Done

File: src/AddVmDialog.js [hunk 23]
```diff
@@ -0,0 +1,310 @@
...
+    $("input[type='text'].combobox").on('change', () => $(this.changeOperatingSystem))
+  }
+
+  submitHandler (e) {
+    e.preventDefault()
+    this.props.type === 'edit' ? this.editVm() : this.createNewVm()
+  }
+
+  getMemory () {
+    return parseInt(this.memory.value) * 1048576
```
Comment (mareklibra) [inline src/AddVmDialog.js]: Done

File: src/AddVmDialog.js [hunk 24]
```diff
@@ -0,0 +1,310 @@
+import React, { PropTypes } from 'react'
```
Comment (mareklibra) [inline src/AddVmDialog.js]: Done

**Comment by mareklibra** [comment]
Replaced by https://github.com/oVirt/ovirt-web-ui/pull/161
